File: 07-trees/00230-Kth_Smallest_Element_In_A_Bst/00230-kth_smallest_element_in_a_bst.go
// 230: Kth Smallest Element In A Bst
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

package main

import (
	"fmt"
)

type TreeNode struct {
    val any
    left *TreeNode
    right *TreeNode
}

type Tree struct {
    root *TreeNode
}

func creator(values []any, root **TreeNode, i, n int) *TreeNode {
    if n==0 {return nil}
    if i<n {
        temp := &TreeNode{values[i], nil, nil}
        *root = temp
        (*root).left = creator(values, &((*root).left), 2*i+1, n);
        (*root).right = creator(values, &((*root).right), 2*i+2, n);
    }
    return *root;
}

func createTree(root **TreeNode, inputs []any) {
    creator(inputs, root, 0, len(inputs))
}

func showTree(root *TreeNode) {
    var q []*TreeNode;
    var result [][]interface{}
    var c []interface{}
    if root==nil { fmt.Println("Empty !"); return; }
    q = append(q, root)
    q = append(q, nil)
    for len(q)!=0 {
        t := q[0]
        q = q[1:]
        if t==nil {
            result = append(result, c)
            c = make([]interface{}, 0)
            if len(q) > 0 {q = append(q, nil)}
        } else {
            c = append(c, t.val)
            if t.left!=nil {q = append(q, t.left)}
            if t.right!=nil {q = append(q, t.right)}
        }
    }

    fmt.Print("["); for _, x := range result {
        fmt.Print("["); for _, y := range x {
            if y==nil { fmt.Print("NULL,"); continue; }
            fmt.Print(y,",")
        }; fmt.Print("\b],")
    }; fmt.Println("\b]")
}

// SOLUTION
var value, result int

func traverse(root *TreeNode) {
	if root==nil || root.val==nil {return}
    traverse(root.left)
    value--
    if value==0 {result = root.val.(int)}
    traverse(root.right)
}

func kthSmallest(root *TreeNode, k int) int {
    value = k
    traverse(root)
    return result
}


func main() {
    tree := Tree{}

    // INPUT
    tn := []any{3,1,4,nil,2}
	k := 1
    createTree(&tree.root, tn)

    // OUTPUT
    result := kthSmallest(tree.root, k)
	fmt.Println(result)
}

File: 07-trees/00230-Kth_Smallest_Element_In_A_Bst/00230-kth_smallest_element_in_a_bst.py
# 230: Kth Smallest Element In A Bst
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/



class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self) -> None:
        self.root = None

    def creator(self, values: list[int], root: TreeNode, i: int, n: int) -> TreeNode:
        if n==0 : return None
        if i<n:
            temp = TreeNode(values[i])
            root = temp
            root.left = self.creator(values, root.left, 2*i+1, n)
            root.right = self.creator(values, root.right, 2*i+2, n)
        return root

    def createTree(self, inputs: list[int]) -> None:
        self.root = self.creator(inputs, self.root, 0, len(inputs))
    
    def showTree(self, root: TreeNode) -> None:
        q = []
        result = []
        c = []
        if root==None : return result
        q.append(root)
        q.append(None)
        while len(q)!=0:
            t = q.pop(0)
            if t==None:
                result.append(c)
                c = []
                if len(q) > 0 : q.append(None)
            else:
                c.append(t.val)
                if t.left!=None : q.append(t.left)
                if t.right!=None : q.append(t.right)

        print("[", end="")
        for x in result:
            if len(x)==0 : continue
            print("[", end="")
            for y in x:
                if y==None: 
                    print("NULL,", end="")
                    continue
                print(y, end=",")
            print("\b]", end=",")
        print("\b]")


class Solution:
    # SOLUTION
    result, value = 0, 0
    
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        def traverse(root: TreeNode) -> None:
            if not root or not root.val : return
            traverse(root.left)
            global value, result
            value-=1
            if value==0 : result = root.val
            traverse(root.right)
        
        global value
        value = k
        traverse(root)
        return result


if __name__ == "__main__":
    o = Solution()
    tree = Tree()

    # INPUT
    tn = [3,1,4,None,2]
    k = 1
    tree.createTree(tn)

    # OUTPUT
    result = o.kthSmallest(tree.root, k)
    print(result)
